lib/xdrgen/generators/python.rb

```diff
@@ -0,0 +1,666 @@
+        when AST::Definitions::Enum
+          render_enum defn
+        when AST::Definitions::Union;
+          render_union defn
+        when AST::Definitions::Typedef
+          render_typedef defn
+        when AST::Definitions::Const
+          render_const defn
+        end
+      end
+
+      def render_const(const)
+        render_const_source_comment @constants_out, const
+        @constants_out.puts "#{const.name}: int = #{const.value}"
+      end
+
+      def render_enum(enum)
+        enum_name = name enum
+        enum_name_underscore = enum_name.underscore
+        @init_out.puts "from .#{enum_name_underscore} import #{enum_name}"
+
+        file_name = "#{enum_name_underscore}.py"
+        out = @output.open(file_name)
+        render_common_import out
+
+        out.puts "__all__ = ['#{enum_name}']"
+
+        out.puts "class #{enum_name}(IntEnum):"
+        out.indent(2) do
+          render_source_comment out, enum
+          enum.members.each do |member|
+            out.puts "#{member.name} = #{member.value}"
+          end
+
+          out.puts <<~HEREDOC
+            def pack(self, packer: Packer) -> None:
+                packer.pack_int(self.value)
+  
+            @classmethod
+            def unpack(cls, unpacker: Unpacker) -> "#{enum_name}":
+                value = unpacker.unpack_int()
+                return cls(value)
+          HEREDOC
+
+          render_xdr_utils out, enum_name
+          out.close
+        end
+      end
+
+      def render_typedef(typedef)
+        typedef_name = typedef.name.camelize
+        typedef_name_underscore = typedef.name.underscore
+
+        @init_out.puts "from .#{typedef_name_underscore} import #{typedef_name}"
+
+        file_name = "#{typedef_name_underscore}.py"
+        out = @output.open(file_name)
+        render_common_import out
+
+        render_import out, typedef, typedef_name
+
+        out.puts "__all__ = ['#{typedef_name}']"
+
+        out.puts "class #{typedef_name}:"
+        out.indent(2) do
+          render_source_comment(out, typedef)
+          out.puts "def __init__(self, #{typedef_name_underscore}: #{type_hint_string typedef, typedef_name}) -> None:"
+          out.indent(2) do
+            render_array_length_checker typedef, out
+            out.puts "self.#{typedef_name_underscore} = #{typedef_name_underscore}"
+          end
+
+          out.puts "def pack(self, packer: Packer) -> None:"
+          out.indent(2) do
+            encode_member typedef, out
+          end
+
+          out.puts "@classmethod"
+          out.puts "def unpack(cls, unpacker: Unpacker) -> \"#{typedef_name}\":"
+          out.indent(2) do
+            decode_member typedef, out
+            out.puts "return cls(#{typedef_name_underscore})"
+          end
+          render_xdr_utils(out, typedef_name)
+          out.puts <<~HEREDOC
+            def __eq__(self, other: object):
+                if not isinstance(other, self.__class__):
+                    return NotImplemented
```

Comment:
is `raise` more appropriate here?

Reply:
We should use `return NotImplemented` here, please refer to [Python Docs](https://docs.python.org/3.11/library/constants.html#NotImplemented).

> A special value which should be returned by the binary special methods (e.g. [__eq__()](https://docs.python.org/3.11/reference/datamodel.html#object.__eq__), [__lt__()](https://docs.python.org/3.11/reference/datamodel.html#object.__lt__), [__add__()](https://docs.python.org/3.11/reference/datamodel.html#object.__add__), [__rsub__()](https://docs.python.org/3.11/reference/datamodel.html#object.__rsub__), etc.) to indicate that the operation is not implemented with respect to the other type; may be returned by the in-place binary special methods (e.g. [__imul__()](https://docs.python.org/3.11/reference/datamodel.html#object.__imul__), [__iand__()](https://docs.python.org/3.11/reference/datamodel.html#object.__iand__), etc.) for the same purpose. It should not be evaluated in a boolean context. NotImplemented is the sole instance of the [types.NotImplementedType](https://docs.python.org/3.11/library/types.html#types.NotImplementedType) type.

Reply:
nice, thank you for the reference! :pray: